src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs

```diff
@@ -1484,14 +1484,10 @@ private static NamedTypeSymbol GetNonDisplayClassContainer(NamedTypeSymbol type)
                 var desiredTypeParameters = candidateSubstitutedSourceType.OriginalDefinition.TypeParameters;
 
                 // We need to use a ThreeState, rather than a bool, because we can't distinguish between
-                // a roslyn lambda that only captures "this" and a dev12 lambda that captures nothing
-                // (neither introduces a display class).  This is unnecessary in the state machine case,
-                // because then "this" is hoisted unconditionally.
-                var isDesiredMethodStatic = hasDisplayClassThis
-                    ? ThreeState.False
-                    : (GeneratedNames.GetKind(candidateSubstitutedSourceType.Name) == GeneratedNameKind.StateMachineType)
-                        ? ThreeState.True
-                        : ThreeState.Unknown;
+                // a Roslyn lambda that only captures "this" and a Dev12 lambda that captures nothing
+                // (neither introduces a display class).  Also, "this" may be optimized away in async
+                // or iterator state machines.
+                var isDesiredMethodStatic = hasDisplayClassThis ? ThreeState.False : ThreeState.Unknown;
```

Comment:
Technically, this no longer needs to be a ThreeState.  I'm not sure what that means for consistency.